runtime/src/ecalls.rs

```diff
@@ -1,58 +1,79 @@
 pub use core::fmt::Write;
 
-// To simplify calling out to the environment, we keep the
-// argument registers intact, and place the function number
-// in s2 (rust will not allow us to use s0 or s1).
+#[cfg(target_arch = "riscv32")]
+mod riscv32 {
+    // To simplify calling out to the environment, we keep the
+    // argument registers intact, and place the function number
+    // in s2 (rust will not allow us to use s0 or s1).
 
-macro_rules! ecall {
+    macro_rules! ecall {
     ($n:literal,$inp1:expr,$inp2:expr,$out:expr) => {
         unsafe {
             core::arch::asm!("ecall", in("s2") $n, in("a1") $inp1, in("a2") $inp2, out("a0") $out)
         }
     }
 }
 
-/// Write a string to the output console (if any).
-pub fn write_log(s: &str) {
-    let mut _out: u32;
-    ecall!(1, s.as_ptr(), s.len(), _out);
-}
+    /// Write a string to the output console (if any).
+    pub fn write_log(s: &str) {
+        let mut _out: u32;
+        ecall!(1, s.as_ptr(), s.len(), _out);
+    }
 
-/// Read a byte from the private input tape
-pub fn read_from_private_input() -> Option<u8> {
-    let inp: u32 = 0;
-    let mut out: u32;
-    ecall!(2, inp, inp, out);
-
-    if out == u32::MAX {
-        None
-    } else {
-        Some(out.to_le_bytes()[0])
-    } // u32::MAX is used a sentinel value that there is nothing (left) on the input tape
-}
+    /// Read a byte from the private input tape
+    pub fn read_from_private_input() -> Option<u8> {
```

Comment:
This needs to be defined on native, for example always return `None`, read from stdin, etc.

Reply:
I would vote that this should `panic` on native.

In the future we could explore having some sort of dev infrastructure that hooks into native run code and lets your replicate (private) input type behavior, but for now I think it's better for it to blow up than to do the wrong thing.

Reply:
Doesn't make much difference to me as long as it's documented. Prohibiting using it on native target with panic sounds worse than returning `None`, forcing people to do
```rust
#[cfg(target_arch = "riscv32")]
let input = read_from_private_input();
#[cfg(not(target_arch = "riscv32"))]
let input = read_from_stdin(); // or None
```

Reply:
I agree it's "worse" in some sense, but if we want to match the developer UX we should do it correctly, rather than have a program differ depending on arch.

This is _really important_, since often a program that is proved on the VM might be audited/formally verified off the VM, so not having non-explicit differences in behavior is an issue for soundness.